src/Apropos/Gen.hs

```diff
@@ -22,17 +27,79 @@ module Apropos.Gen (
   linear,
   singleton,
 ) where
-
 import Apropos.Gen.Range
-import Control.Monad (replicateM, void)
+import Control.Monad (replicateM,(>=>))
 import Control.Monad.Free
 import Control.Monad.Trans (lift)
 import Control.Monad.Trans.Except (ExceptT, runExceptT, throwE)
+import Control.Monad.Trans.Writer (WriterT, runWriterT, tell)
 import Data.String (fromString)
-import Hedgehog (Property, PropertyT)
+import Hedgehog (PropertyT)
 import Hedgehog qualified as H
 import Hedgehog.Gen qualified as HGen
 import Hedgehog.Range qualified as HRange
+import Data.Set (Set)
+import Data.Set qualified as Set
+
+
+forAllWithRetries :: Show a => Int -> Gen a -> PropertyT IO a
+forAllWithRetries lim g = do
+  (ee,labels) <- H.forAll $ runWriterT (runExceptT $ handleRetries lim g)
+  mapM_ (H.label . fromString) labels
+  case ee of
+    Left Retry -> H.footnote "retry limit reached" >> H.failure
+    Left (GenException err) -> H.footnote err >> H.failure
+    Right a -> pure a
+
+
+forAllNoShrink :: Show a => Gen a -> PropertyT IO a
+forAllNoShrink g = do
+  (ee,labels) <- H.forAll $ HGen.prune $ runWriterT (runExceptT $ gen g)
+  mapM_ (H.label . fromString) labels
+  case ee of
+    Left Retry -> H.footnote "retry limit reached" >> H.failure
+    Left (GenException err) -> H.footnote err >> H.failure
+    Right a -> pure a
+
+forAll :: Show a => Gen a -> PropertyT IO a
+forAll g = do
+  (ee,labels) <- H.forAll $ runWriterT (runExceptT $ gen g)
+  mapM_ (H.label . fromString) labels
+  case ee of
+    Left Retry -> H.footnote "retry limit reached" >> H.failure
+    Left (GenException err) -> H.footnote err >> H.failure
+    Right a -> pure a
+
+data Morph m where
+  Source :: Gen m -> Morph m
+  Morph :: Morph m -> (m -> Gen [m -> Gen m]) -> Morph m
+
+--TODO use Morphism instead then we can print the Morphism name
+--i.e. Morph :: Morph p m -> (m -> Gen [Morphism p m]) -> Morph p m
+--also Source has been introduced - we want to be able to partition the space with a set of sources! XD
+instance Show (m -> Gen m) where
+  show _ = "GenMorphism"
```

Comment:
This lets us inject quantification between each morphism which contains the shrink tree to operate within the scope of the Morphism AST. The shrinker was eating all my RAM when it was trying to operate on the composition of all the morphisms (I suspect this is because it was trying to shrink the list of morphisms - which has no reasonable shrink behaviour AFAICT).